### PR TITLE
Added auto-completion for shells

### PR DESCRIPTION
**Pull request by TheRustyPickle**
#63 Added auto-completion support (Linux only) for Bash, Zsh, and Fish shells on Tab press. Gets initialized at the start of the program run

Screenshot from 2023-02-28 04-59-30
Screenshot from 2023-02-28 05-00-06
Screenshot from 2023-02-28 05-00-26

### CODE REVIEW

**Comment by 24seconds** [comment]
[Question]
I also want to test the auto completion. But.. I have no idea how to do that.
Could you provide some detail guide so that I can reproduce like the image you attach?

**Comment by TheRustyPickle** [comment]
@24seconds running the app once will add autocompletion for the shell which was used.

File: src/autocomplete.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
+use clap::Command;
+use clap_complete::{generate, Generator, Shell};
```
[the remark this answers is not in the document]
Comment (24seconds) [reply src/autocomplete.rs:2]: > running the app once will add autocompletion for the shell which was used
Oh then.. should we use `generate_to`? 

As I remember, whenever we install the `git` it's autocompletion works without running once.

What do you think?

**Review by 24seconds — COMMENTED** [review]
Thank you for the work! I'm still reading some material related to `clap_autocomplete` and how `autocompletion works` for the review.

Here is the in the middle of review results. Please take a look 🙇

File: src/autocomplete.rs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+        // convert to readable version
+        let stdout = String::from_utf8(output.stdout)?;
+        let shell = stdout.trim();
+
+        match shell {
+            "bash" => Ok(Shell::Bash),
+            "zsh" => Ok(Shell::Zsh),
+            "fish" => Ok(Shell::Fish),
+            //"elvish" => Ok(Shell::Elvish),
```
[the remark this answers is not in the document]
Comment (TheRustyPickle) [reply src/autocomplete.rs:89]: I put it in a comment in case you or anyone else want to add autocomplete support for elvish. The Elvish shell autocomplete process seemed too complicated for me. I can remove the lines if you want me to.

[the remark this answers is not in the document]
Comment (24seconds) [reply src/autocomplete.rs:89]: Aha then please remove this line.

File: Cargo.toml [hunk 3]
```diff
@@ -32,6 +32,7 @@ serde = "1.0.144"
 reqwest = { version = "0.11", features = ["json"] }
 colored = "2"
 bincode = { version = "2.0.0-rc.1", features = ["alloc"]}
+clap_complete = "3.2.2"
```
[the remark this answers is not in the document]
Comment (TheRustyPickle) [reply Cargo.toml:35]: Correct, it's related to the clap's version. If I use the latest version I get some kind of error with generate() which I don't understand. Something to do with the clap's main command generation

File: src/autocomplete.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheRustyPickle) [reply src/autocomplete.rs:2]: generate_to is for generating the autocomplete script during compile time instead of runtime with generate. It may enable initiating the autocomplete if `cargo install` or `cargo build` is used. Further research is needed before I can comment further.

[the remark this answers is not in the document]
Comment (TheRustyPickle) [reply src/autocomplete.rs:2]: @24seconds did some testing with `generate_to`. Check the example with `src/cli.rs` [here](https://docs.rs/clap_complete/latest/clap_complete/generator/fn.generate_to.html). 

`get_main_command` from `application.rs` does a lot of work from other rs files. I can't get the build.rs to compile because of the `use crate` imports inside `application.rs`. It is likely the process needs to be simplified before this can work.

**Comment by 24seconds** [comment]
Thank you for handling the comment and leave comments. I didn't check the comment yet.. 
This week I'm quite busy. I will take a look on Friday night or Saturday after noon (KST). @TheRustyPickle

File: src/autocomplete.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (24seconds) [reply src/autocomplete.rs:2]: Got it. For this PR, let's use `generate`

File: src/autocomplete.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    cmd: &mut Command,
+    file_name: &str,
+) -> std::io::Result<()> {
+    let mut file = File::create(file_name)?;
+    generate(gen, cmd, cmd.get_name().to_string(), &mut file);
+    Ok(())
+}
+
+/// Adds auto completion for the shell that is being used
+pub fn add_autocomplete() -> Result<(), Box<dyn std::error::Error>> {
```
Comment (24seconds) [inline src/autocomplete.rs:23]: [Suggestion]
For the better error message, how about define custom error for auto completion?  
`rust-cli-pomodoro` defines many custom errors in `error.rs` file.

Link: https://github.com/24seconds/rust-cli-pomodoro/blob/0ab2e16ba16cd2fdc9c8777d878bbf19132047f0/src/error.rs#L1

File: src/autocomplete.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    let current_shell = get_current_shell();
+
+    if let Ok(shell) = current_shell {
+        debug!("Current shell: {}", shell);
+        let mut main_command = get_main_command();
+        let (file_name, mut moving_path) = match shell {
+            Shell::Fish => ("pomodoro.fish", validate_path(Shell::Fish)?),
+            Shell::Zsh => ("_pomodoro.zsh", validate_path(Shell::Zsh)?),
```
Comment (24seconds) [inline src/autocomplete.rs:31]: [Question]
Is there a reason to use `_pomodoro.zsh`, not `pomodoro.zsh` for zsh Shell?

File: src/autocomplete.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        fs::copy(file_name, moving_path)?;
+        fs::remove_file(file_name)?;
+        edit_shell_file(shell)?;
+    }
+    Ok(())
+}
+
+/// Returns the current shell
+fn get_current_shell() -> Result<Shell, Box<dyn std::error::Error>> {
+    if cfg!(target_os = "linux") {
```
Comment (24seconds) [inline src/autocomplete.rs:57]: [Question]
Is there a reason to target only linux? How about mac or window?

**Review by 24seconds — COMMENTED** [review]
I'm still understanding the auto completion.. 🙇 . I left some comments.

**Comment by 24seconds** [comment]
### How completion works
So far I looked up the `tab completion` or `autocompletion`. And how `completion` works can be summarized below.
- for the cli application, we need to generate `completion script`. For example, in zsh shell, the usuall `completion script` filename should be `_your_application_name`. In our case, it's `_pomodoro`.
- And we should put the `completion script` to proper path so that the zsh shell can figure out. It seems that the `completion script`  should be placed in `$fpath`.
    - zsh completion explanation link: https://github.com/zsh-users/zsh-completions/blob/98ea8e685c1a9630a21b39a1596cb94794d1cb3c/zsh-completions-howto.org?plain=1#L31
- Other shells work same.

### Suggestion
I found good example of `completion`. It's [rustup](https://github.com/rust-lang/rustup/tree/master). rustup supoprts `completion` command. You can run `rustup completions --help` and see the contents. Or check the installation doc link - https://github.com/rust-lang/rustup/blob/affbdfe0a37d92444e7c3c111043cd336e2f7728/doc/src/installation/index.md?plain=1#L68.

In this PR, the code tries to do two things.
- create `completion script`
- move `completion script` to proper path

I think we should approach similar to rustup. Let `completion` be optional. `rust-cli-pomodoro` only provides `completion script` generation command like rustup. And with the ease of adoption, we can write kind guidance.
Another good example of guidance is `nvm` I guess. It provides code snippet for shell profile file - link: https://github.com/nvm-sh/nvm/blob/edacf8275e3bef4a80971625ed89df13a9af427c/README.md?plain=1#L240.

To sum up, I suggest
- provide `completions` command very similar to `rustup`
- write guidance - (I'm not sure which guidance is preferred way - `rustup` or `nvm`?)

What do you think? @TheRustyPickle

File: src/autocomplete.rs (hunk 6 above)
Comment (TheRustyPickle) [reply src/autocomplete.rs:57]: I don't have a mac or windows. I don't know if the process would be the same over there.

**Comment by TheRustyPickle** [comment]
Sounds good @24seconds. I liked the `rustup`'s approach better and I think is easier to understand. If I understood you right, the new approach will have a command like this instead of detecting the shell manually:

`pomodoro completions fish > path`

where `generate` will return a `stdout`, reducing most of the complexity. If I have the green light I can try working on this.

**Comment by 24seconds** [comment]
You are right. Thank you @TheRustyPickle !

**Comment by TheRustyPickle** [comment]
Thank you @24seconds! I'm not yet very familiar with clap yet but will try to commit the new changes as soon as I am able.

**Comment by TheRustyPickle** [comment]
Hello @24seconds, correct me here if I am wrong here. 

If I don't enter the pomodoro cli interface by typing `pomodoro` or `cargo run` and try to pass args like this `pomodoro create -d`, it tries to connect to/create UDS. 

If we want to use `pomodoro completions fish > path` or how it was done in `rustup` and use the `stdout`, it needs to be done without entering the pomodoro cli interface. 

I'm not sure if it was expected or whether this side is incomplete, I'm getting this error:

`Error: Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }`

Kindly advise on how to proceed here. I'm quite confused and not sure how things are working on the UDS side. Thank you for your time.

**Comment by 24seconds** [comment]
I understand the situation.

Ah.. `rust-cli-pomodoro` architecture is quite unusual. There are two modes.  
For more information, check this section - https://github.com/24seconds/rust-cli-pomodoro/blob/0ab2e16ba16cd2fdc9c8777d878bbf19132047f0/README.md?plain=1#L24
1. Single mode: Run `rust-cli-pomodoro` by typing the `pomodoro` command. Then the app is running in terminal and you can type some commands **in that terminal**.
2. Communication mode: I found that it's burdensome for me - whenever I type, I need to go to that terminal. Usually I opened multiple terminal tabs. That is why `ipc` is used. If we type command in another terminal, then the already running `rust-cli-pomodoro` listens and serves command. (It works like client-server)

Therefore, to determine which mode it is, we have a function named `detect_command_type`.  
Link: https://github.com/24seconds/rust-cli-pomodoro/blob/0ab2e16ba16cd2fdc9c8777d878bbf19132047f0/src/main.rs#L162
Currently there are two `command type` - StarUP and UdsClient. Now I guess it's time to add one more command type - `completion` ? `autocomplete`? something like this (please suggest command naming). 

Suggestion
- refactor `detect_command_type` so that we can distinguish which command type it is. Currently `detect_comamnd_type` naively determines command type
- Write some login in main.rs depending on `detect_command_type` result.

Then you can achieve what you wanted I guess. Please ask question @TheRustyPickle  if something is unclear 🙇

**Comment by TheRustyPickle** [comment]
Thank you for the kind explanation, @24seconds! It is much clearer for me and I have some ideas on what to work on now🙏

**Comment by TheRustyPickle** [comment]
@24seconds when you have the time, please take a look at the new changes and let me know your thoughts. Thank you!

if you already have `cargo install rust-cli-pomodoro`, you should be able to do this

`pomodoro completion fish > pomodoro.fish`.

**Review by 24seconds — APPROVED** [review]
Looks good to me!! Super!

Thank you for the patient and long long communication. Because of you, I can understand this feature and `rust-cli-pomodoro` now has tab completion!

I will update version soon. Again, thank you!

**Comment by 24seconds** [comment]
image

Confirm that it works well in zsh, macOS

**Comment by 24seconds** [comment]
[next step]
For the general usage, maybe `rust-cli-pomodoro` needs guidance like rustup. Let me work on it if I'm available.

**Comment by TheRustyPickle** [comment]
Thank you so much for being so helpful and patient throughout the process, @24seconds! 🙏
